module/zfs/dnode.c

```diff
@@ -1565,6 +1570,8 @@ dnode_hold_impl(objset_t *os, uint64_t object, int flag, int slots,
 			} else {
 				dn = dnode_create(os, dn_block + idx, db,
 				    object, dnh);
+				dmu_buf_add_user_size(&db->db,
+				    sizeof (dnode_t));
```

Comment:
Moving the `dmu_buf_add/sub_user_size()` calls in to `dnode_create/destroy()` looks like it would help simply things and make sure no call locations are missed.

Reply:
We can't do that on the remove side, because when `dnode_destroy()` is called from the evict side, the dbuf may have already been, and that's the only path back into the userdata. That's why `dbuf_evict_user()` adjusts the cache ahead of time.

On the create side, the patch is:

```
diff --git module/zfs/dnode.c module/zfs/dnode.c
index 029d9df8a..b36ee578a 100644
--- module/zfs/dnode.c
+++ module/zfs/dnode.c
@@ -625,6 +625,8 @@ dnode_create(objset_t *os, dnode_phys_t *dnp, dmu_buf_impl_t *db,
        mutex_exit(&os->os_lock);

        arc_space_consume(sizeof (dnode_t), ARC_SPACE_DNODE);
+       if (db != NULL)
+               dmu_buf_add_user_size(&db->db, sizeof (dnode_t));

        return (dn);
 }
```

Its still conditional because of the `dnode_special_open()` case, which has no dbuf.

So I think I prefer it the way it is, where the caller has to do the accounting for both cases. But I'm not strongly opinionated and it is all weird either way :)

Reply:
Well that's unfortunate, then I'm good with this.  Thanks for the explanation.